### ISSUE TITLE
Retrieve data as bytes not as string

### ISSUE THREAD

**Issue by zmic** [jim-easterbrook/python-exiv2#6]
Hi Jim,
I wrote a bytes object to the image file but I can't see an obvious way to retrieve the data back as bytes. I know there's a toString() function but that returns a string.

```
def reload(path):
    img = exiv2.ImageFactory.open(path)
    img.readMetadata()
    data = img.iptcData()["Iptc.Application2.Caption"]
    data = data.value()
    ??? how do I get the original bytes object ???
```

**Comment by jim-easterbrook** [comment]
The conversion to `str` is done within the SWIG generated wrapper using a `utf-8` codec with `surrogateescape` error handling. This allows you to get the original raw `bytes` like this:
```
value = datum.toString()
raw_value = value.encode('utf-8', errors='surrogateescape')
```

**Comment by zmic** [comment]
Doesn't look very kosher to me since I never encoded anything in utf-8... the data is a dump of a pytorch tensor. That surragoteescape will ensure a 1 to 1 with the original?

**Comment by jim-easterbrook** [comment]
The SWIG interface is generated from the Exiv2 API which uses C++ `std::string` for data values. The image metadata itself could be in any text encoding, or plain ascii, there's no way to know. (I've got a collection of rogue files with all sorts of rubbish in them.) SWIG always converts ` std::string` to a Python `str` using `utf-8`. Fortunately the `surragoteescape` error handler gives us a string that can be converted back to the original bytes (1 to 1) and no exception is raised.

**Comment by jim-easterbrook** [comment]
PS It's against the metadata standards (Exif, IPTC, and XMP) to write binary data to their text fields.

**Comment by zmic** [comment]
Sure but 1. some fields are binary and 2. Exif and IPTC most certainly don't assume that you will use utf-8 encoding. Since you want to offer a low-level interface to exiv2, and the original interface exposes char*, it seems totally natural that your wrapper should offer setting and getting Python bytes objects so the user can choose how to encode/decode their data rather than be surprised by implicit conversions or unnecessary go-arounds.

I'm not familiar with SWIG but surely there is some way to cajole SWIG to allow for byte objects when the original functions want char*?

Thank you for your quick responses!

**Comment by jim-easterbrook** [comment]
The API also uses `std::string` for things we actually want to be strings, so it would be a lot of work (if it's even possible) to make the interface do different things in different places. And all new image files should have string data utf-8 encoded, according to the standards. `char*` is not used in the API as far as I know.
Binary fields are mostly available as the appropriate data type.

**Comment by zmic** [comment]
Yes but std::string obviously encapsulates char*, there're not std::wstrings.

I decided to use a binary field instead (Iptc.Application2.Preview) which can contain binary data but I'm having trouble setting the value.

```
def write_file():

    rand_tensor = torch.rand(1, 4, 128, 128)
    b = BytesIO()
    torch.save(rand_tensor, b)
    preview_data = b.getvalue()
    print(len(preview_data))

    img = Image.open(r"H:\GEN\TG\XK\ts_1590541947676_41_gen_r2.jpg")
    temp = BytesIO()
    img.save(temp, format="jpeg",quality=90)
    jpg_bytes = temp.getvalue()
    img = exiv2.ImageFactory.open(jpg_bytes)
    img.readMetadata()
    new_iptc_data = exiv2.IptcData()
    **new_iptc_data["Iptc.Application2.Preview"].value = preview_data**
    img.setIptcData(new_iptc_data)
    img.setIptcData(new_iptc_data)
    img.writeMetadata()
    
    path = "test0.jpg"    
    open(path, 'wb').write(bytes(img.io()))
```

But this gives error when the metadata is written:

```
Traceback (most recent call last):
  File "H:\dev\DU\DU\scripts\test_exiv.py", line 42, in <module>
    write_file()
  File "H:\dev\DU\DU\scripts\test_exiv.py", line 28, in write_file
    img.writeMetadata()
exiv2.Exiv2Error: Value not set
```

**Comment by zmic** [comment]
I appreciate your efforts to invent a workaround, but I simply want to use the library the way it's intended to use. I find no way to set binary data (bytes object) to new_iptc_data["Iptc.Application2.Preview"]. I don't want to do silly conversions to Unicode strings and back. The field expects image data, not an utf-8 encoded string. If your wrapper doesn't allow it then your wrapper is broken and you need to fix it.

**Comment by jim-easterbrook** [comment]
Try creating an `exiv2.DataValue` with your binary data and setting the iptc datum's value to it. You should be able to construct an instance from any Python object that exposes the buffer interface.

**Comment by jim-easterbrook** [comment]
> You should be able to construct an instance from any Python object that exposes the buffer interface.

I've just been trying this out for myself (I've not needed to use `DataValue` before) and it's not working as it should. Give me a day or two to investigate further.

**Comment by jim-easterbrook** [comment]
Commit 95e22a2 allows passing binary data (e.g. `bytes`) to the `DataValue` constructor and `read()` methods, and passing a writeable buffer (e.g. `bytearray`) to the `copy()` method. I'll try and get a new release out on Monday.

**Comment by zmic** [comment]
That would be awesome, thanks!!

**Comment by jim-easterbrook** [comment]
I've found some unintended side effects of my changes which need to be dealt with. I may need a little longer to get it to a releasable state.

**Comment by zmic** [comment]
Thanks for the quick action, Jim!

**Comment by jim-easterbrook** [comment]
Newly released v0.13.1 adds a `data()` method to `Exiv2::StringValueBase` and its derived types. This should give you copy-free access to the raw bytes of strings and comments.
